Add refs/decomps for dot/vdot

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108194
* #108188

Follow-up on https://github.com/pytorch/pytorch/issues/108127#issuecomment-1698142427

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

lezcano committed:
Add refs/decomps for dot/vdot

Follow-up on https://github.com/pytorch/pytorch/issues/108127#issuecomment-1698142427

[ghstack-poisoned]

peterbell10 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1168,15 +1168,10 @@ def make_arg_conj(size):
                      error_regex='inconsistent tensor size')
-    if device != "cpu":
-        ErrorInput(SampleInput(make_input(3), args=(make_input(3, device="cpu"),)),
-                   error_regex='Expected all tensors to be on the same device')
```
Why are we changing the tests here? Shouldn't you fix the refs to match the test.

peterbell10 replied:
Adding more functionality to the ref seems fine, but removing testing from eager definitely seems bad,

peterbell10 replied:
Could you override the `error_inputs_func` of the ref info to pass in an `is_ref=True` option?

lezcano replied:
Fair enough. Done.

peterbell10 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -20957,6 +20952,26 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
+        "_refs.dot",
+        torch_opinfo_name="dot",
+        # .conj() returns a view on complex tensors??
+        validate_view_consistency=False,
```
Why aren't we getting the view behavior right here?

peterbell10 replied:
At a guess, `return torch.dot(self, other.conj()).conj()` probably needs to come outside of `elementwise_type_promotion_wrapper`.

lezcano replied:
Not quite. I checked under this branch:
```python
    elif other.is_conj():
        return torch.dot(self, other.conj()).conj()
```
which is the one that triggers the issue, and `torch.dot(self, other.conj())` is not a view, but `torch.dot(self, other.conj()).conj()` is a view.
On the other hand, if I put the same code in ATen, it says that `torch.dot(self, other.conj()).conj()` is conj but not a view... I'm submitting a different fix.

lezcano replied:
Nvm, the other fix doesn't work. I'm not quite sure what's going on there.
I'm voting for leaving as is right now as we don't even support complex numbers at the moment in inductor.

peterbell10 replied:
Okay the eager result having `is_conj()` but not `_is_view()` is weird but not really our fault here, so I guess thats fine. This comment should be updated to reflect that though; `.conj()` is expected to return a view so the comment doesn't make much sense.

lezcano committed:
Update on "Add refs/decomps for dot/vdot"


Follow-up on https://github.com/pytorch/pytorch/issues/108127#issuecomment-1698142427

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

lezcano committed:
Update on "Add refs/decomps for dot/vdot"


Follow-up on https://github.com/pytorch/pytorch/issues/108127#issuecomment-1698142427

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

lezcano committed:
Update on "Add refs/decomps for dot/vdot"


Follow-up on https://github.com/pytorch/pytorch/issues/108127#issuecomment-1698142427

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

lezcano committed:
Update on "Add refs/decomps for dot/vdot"


Follow-up on https://github.com/pytorch/pytorch/issues/108127#issuecomment-1698142427

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

lezcano commented:
@pytorchbot merge

lezcano committed:
Update on "Add refs/decomps for dot/vdot"


Follow-up on https://github.com/pytorch/pytorch/issues/108127#issuecomment-1698142427

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

lezcano commented on aten/src/ATen/native/cuda/Blas.cpp:
```diff
@@ -496,12 +496,6 @@ inline void dot_check(const Tensor& self, const Tensor& other) {
       " and ",
       other.numel(),
       " elements respectively");
-  TORCH_CHECK(
-      self.device() == other.device(),
-      "expected all tensors to be on the same device. Found: ",
-      self.device(),
-      ", ",
-      other.device());
```
@peterbell10 removing this as it's unreachable. We currently check this in the wrapper and throw
```
Expected all tensors to be on the same device, but found at least two devices, cpu and cuda:0! (when checking argument for argument tensor in method wrapper_CUDA__dot
```

lezcano commented:
@pytorchbot merge
